File: .config/nvim/lua/plugins/lsp.lua
local util = require('utils')
local on_attach = function(client, bufnr)
  local opts = { silent=true }
  util.bufmap(bufnr, "n", "<Leader>lD", "<cmd>lua vim.lsp.buf.declaration()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>ld", "<cmd>lua vim.lsp.buf.definition()<CR>", opts)
  util.bufmap(bufnr, "n", "K", "<cmd>lua vim.lsp.buf.hover()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>li", "<cmd>lua vim.lsp.buf.implementation()<CR>", opts)
  util.bufmap(bufnr, "n", "<C-k>", "<cmd>lua vim.lsp.buf.signature_help()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>lwa", "<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>lwr", "<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>lwl", "<cmd>lua print(vim.inspect(vim.lsp.buf.list_workspace_folders()))<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>ltd", "<cmd>lua vim.lsp.buf.type_definition()<CR>", opts)
  util.bufmap(bufnr, "n", "<F2>", "<cmd>lua vim.lsp.buf.rename()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>lca", "<cmd>lua vim.lsp.buf.code_action()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>lr", "<cmd>lua vim.lsp.buf.references()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>le", "<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>", opts)
  util.bufmap(bufnr, "n", "[d", "<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>", opts)
  util.bufmap(bufnr, "n", "]d", "<cmd>lua vim.lsp.diagnostic.goto_next()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>lq", "<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>", opts)
  util.bufmap(bufnr, "n", "<Leader>lf", "<cmd>lua vim.lsp.buf.formatting()<CR>", opts)
end

local lsp_installer = require("nvim-lsp-installer")
lsp_installer.on_server_ready(function(server)
  local opts = {}
  opts.on_attach = on_attach

  server:setup(opts)
  vim.cmd [[ do User LspAttachBuffers ]]
  opts.capabilities = require('cmp_nvim_lsp').update_capabilities(vim.lsp.protocol.make_client_capabilities())
end)

util.opt.completeopt = 'menu,menuone,noselect'

File: denovim/denops/config/plugins/lsp.ts
import type { Plug } from "https://deno.land/x/dvpm@1.0.0/mod.ts";
import { execute } from "https://deno.land/x/denops_std@v5.0.0/helper/execute.ts";
import { notify } from "../util.ts";

export const lsp: Plug[] = [
  {
    url: "neovim/nvim-lspconfig",
    dependencies: [
      { url: "williamboman/mason.nvim" },
      { url: "williamboman/mason-lspconfig.nvim" },
      { url: "folke/trouble.nvim" },
      // { url: "lukas-reineke/lsp-format.nvim" },
      { url: "jose-elias-alvarez/null-ls.nvim" },
      {
        url: "simrat39/symbols-outline.nvim",
        after: async ({ denops }) => {
          await denops.cmd(`lua require("symbols-outline").setup()`);
        },
      },
      {
        url: "SmiteshP/nvim-navic",
        after: async ({ denops }) => {
          await denops.call(
            `luaeval`,
            `require("nvim-navic").setup(_A.param)`,
            {
              param: { lsp: { auto_attach: true }, highlight: true },
            },
          );
        },
      },
      {
        url: "folke/neodev.nvim",
        dependencies: [{ url: "neovim/nvim-lspconfig" }],
        after: async ({ denops }) => {
          await denops.call(
            `luaeval`,
            `require("neodev").setup(_A.param)`,
            {
              param: { experimental: { pathStrict: true } },
            },
          );
        },
      },
      {
        url: "folke/neoconf.nvim",
        dependencies: [{ url: "neovim/nvim-lspconfig" }],
        after: async ({ denops }) => {
          await denops.cmd(`lua require("neoconf").setup()`);
        },
      },
    ],

    after: async ({ denops }) => {
      await execute(
        denops,
        `
lua << EOB
require("mason").setup({
            providers = {
              "mason.providers.client",
              "mason.providers.registry-api",
            },
})
ensure_installed = {
            "stylua",
            "shellcheck",
            "shfmt",
            "flake8",
            "prettier",
            "lua-language-server",
            "vim-language-server",
            "deno",
          }
require("mason-lspconfig").setup()
for _, tool in ipairs(ensure_installed) do
  p = require("mason-registry").get_package(tool)
  if not p:is_installed() then
    p:install()
  end
end


function on_attach(on_attach)
  vim.api.nvim_create_autocmd("LspAttach", {
    callback = function(args)
      bufnr = args.buf
      client = vim.lsp.get_client_by_id(args.data.client_id)
      on_attach(client, bufnr)
    end,
  })
end

on_attach(function(client, bufnr)
  opts = { noremap = true, silent = true, buffer = bufnr }
  vim.keymap.set("n", "g[", function() vim.diagnostic.goto_prev() end, opts)
  vim.keymap.set("n", "g]", function() vim.diagnostic.gotio_next() end, opts)
  vim.keymap.set("n", "gD", function() vim.lsp.buf.declaration() end, opts)
  vim.keymap.set("n", "gd", function() vim.lsp.buf.definition() end, opts)
  vim.keymap.set("n", "K", function() vim.lsp.buf.hover() end, opts)
  vim.keymap.set("n", "gi", function() vim.lsp.buf.implementation() end, opts)
  vim.keymap.set("n", "gK", function() vim.lsp.buf.signature_help() end, opts)
  vim.keymap.set("n", "gr", function() vim.lsp.buf.rename() end, opts)
  vim.keymap.set("n", "ga", function() vim.lsp.buf.code_action() end, opts)
  vim.keymap.set("n", "gR", function() vim.lsp.buf.references() end, opts)
  vim.keymap.set("n", "<Leader>F", function() vim.lsp.buf.format() end, opts)
end)

lspconfig = require("lspconfig")
capabilities = vim.lsp.protocol.make_client_capabilities()
capabilities.textDocument.foldingRange = {
            dynamicRegistration = false,
            lineFoldingOnly = true,
          }
capabilities.textDocument.completion.completionItem.snippetSupport = true
capabilities.textDocument.completion.completionItem.resolveSupport = {
  properties = {
    "documentation",
    "detail",
    "additionalTextEdits"
  }
}
options = {
  capabilities = capabilities,
}

require("mason-lspconfig").setup_handlers({
function(server_name)
  lspconfig[server_name].setup(options)
end,

lua_ls = function() 
  lspconfig["lua_ls"].setup(vim.tbl_deep_extend("force", options, {
    settings = {
      Lua = {
        completion = {
          callSnippet = "Replace",
        },
      },
    },
  }))
end,

denols = function()
  lspconfig["denols"].setup(vim.tbl_deep_extend("force", options, {
    root_dir = lspconfig.util.root_pattern("deno.json"),
    init_options = {
      lint = true,
      unstable = true,
      suggest = {
        imports = {
          hosts = {
            ["https://deno.land"] = true,
            ["https://cdn.nest.land"] = true,
            ["https://crux.land"] = true,
          },
        },
      },
    },
  }))
end,
kotlin_language_server = function()
  lspconfig["kotlin_language_server"].setup(vim.tbl_deep_extend("force", options, {
    -- root_dir = lspconfig.util.root_pattern("build.gradle.kts"),
  }))
end,

powershell_es = function()
  lspconfig["powershell_es"].setup(vim.tbl_deep_extend("force", options, {
  mason = false,
  settings = {
    powershell = {
      codeFormatting = {
        preset = "OTBS",
        newLineAfterOpenBrace = false,
        newLineAfterCloseBrace = true,
      },
    },
  },
  on_attach = on_attach(function(client, bufnr)
    client.server_capabilities.semanticTokensProvider = nil
  end),
}))
end,
})`,
      );
      await notify(denops, "lsp loaded");
    },
  },
];
